Update test descriptions of functionality test suite 4

fluiddot opened the pull request:
This PR is a follow up on [this comment](https://github.com/wordpress-mobile/gutenberg-mobile/pull/5977#discussion_r1279801896) regarding updating the test descriptions.

**To test:**
* Observe that all CI jobs (including the optional e2e tests) pass.

PR submission checklist:

- [x] I have considered adding unit tests where possible.
- [ ] I have considered if this change warrants user-facing release notes [more info](https://github.com/wordpress-mobile/gutenberg-mobile/blob/trunk/Release-notes.md) and have added them to [RELEASE-NOTES.txt](https://github.com/wordpress-mobile/gutenberg-mobile/blob/trunk/RELEASE-NOTES.txt) if necessary.

fluiddot committed:
Update test descriptions of functionality test suite 4

geriux commented on __device-tests__/image-snapshots/gutenberg-editor-sanity-test-4-visual-test-js-gutenberg-editor-test-suite-4-buttons-block-should-render-dark-mode-colors-1-ios.png:
I see a red underline below the text, I guess because of the check spelling being enabled on your simulator, I wonder if this would cause issues on CI since I believe the test devices we use don't have autocorrect enabled. Have you encountered any flakiness about it?

fluiddot replied:
Ah, good point. I checked the failures and doesn't seem to be related to visual checks. In any case, I'll review the ones I generated and disable the spelling just in case. Thanks for the feedback 🙇 !

fluiddot replied:
This was solved in https://github.com/wordpress-mobile/gutenberg-mobile/pull/6013/commits/1680eb635b861ceb405850dbe0178e36faf0f2e1.

fluiddot replied:
As I shared in [this comment](https://github.com/wordpress-mobile/gutenberg-mobile/pull/6013/files#r1281551891), this error was caused by using a non-English keyboard, so it's unrelated to having the spell-checking option enabled. In fact, CI's simulator has spell-checking enabled when running the tests.

fluiddot replied:
In relation to this, I created the https://github.com/wordpress-mobile/gutenberg-mobile/pull/6038 PR to incorporate a new tool that will be used here to solve the issue. @geriux let me know if you could take a look, thanks 🙇 !

fluiddot committed:
Remove spell-check visual hints from e2e screenshots

fluiddot commented:
@dcalhoun @geriux the PR is ready for review, thanks 🙇 !

dcalhoun reviewed (approved):
🚀

fluiddot commented on __device-tests__/image-snapshots/gutenberg-editor-sanity-test-4-visual-test-js-gutenberg-editor-test-suite-4-buttons-block-should-have-its-selection-caret-with-the-same-color-as-the-font-1-ios.png:
This screenshot doesn't match the one generated on CI. It expects a `Replace` option in the context menu, however, that only happens when enabling spell-checking 🤔. I originally thought that we needed to use a simulator with spell-checking disabled, but seems that's incorrect. The error I tried to fix in relation to this comment was actually caused by using a keyboard set to a non-English language, not related to spell-checking.

![image](https://github.com/wordpress-mobile/gutenberg-mobile/assets/14905380/cc6d4b5e-d4fb-4843-980e-04d6e77f5b52)

fluiddot replied:
For some reason, I'm not getting the context menu with the `Replace` option when running the e2e locally. However, the menu was correct when following the same steps manually 🤔. I finally had to update the test to ensure the context menu displays the expected options (https://github.com/wordpress-mobile/gutenberg-mobile/pull/6013/commits/1989eca6bc33e93fced1b84f92590b5023aae581).

The screenshot has been updated in https://github.com/wordpress-mobile/gutenberg-mobile/pull/6013/commits/b255282acb184b7cda56a8f7c8a82d984648fc09.

fluiddot replied:
This has been finally solved via https://github.com/wordpress-mobile/gutenberg-mobile/pull/6013/commits/f5626013205bf31eba2bfa9057a13e86e01c1fdb.

fluiddot committed:
Fix Buttons block e2e test

On iOS, the context menu displayed when selecting the text wasn't displaying the expected options. To solve this the text is clicked before being selected. Additionally, previous actions for selecting the Button block have been removed as aren't necessary.

fluiddot committed:
Update Buttons block e2e test screenshot

fluiddot committed:
Add extra wait in Buttons block e2e test

fluiddot commented:
I'm afraid the visual test `Buttons block - should have its selection / caret with the same color as the font` still fails despite the recent fixes. 

![download-1](https://github.com/wordpress-mobile/gutenberg-mobile/assets/14905380/514eab5d-4398-4670-a833-90bd624cf45e)

I'll keep investigating and use a separate PR for testing with CI, as seems it's likely to have discrepancies in the context menu.

dcalhoun commented:
> I'm afraid the visual test `Buttons block - should have its selection / caret with the same color as the font` still fails despite the recent fixes.

My guess is that the skinnier, translucent context menu is a result of the reveal animation having not yet finished. Is this issue resolved if a timeout is added to await the animation?

fluiddot commented:
> > I'm afraid the visual test `Buttons block - should have its selection / caret with the same color as the font` still fails despite the recent fixes.
> 
> My guess is that the skinnier, translucent context menu is a result of the reveal animation having not yet finished. Is this issue resolved if a timeout is added to await the animation?

Yep, I thought the same, in fact, I tried to solve it by adding an extra wait in https://github.com/wordpress-mobile/gutenberg-mobile/pull/6013/commits/a70d9c7a8ea2508989ae49d143fa8cc871fe7021. But still, it didn't work. Maybe it can be fixed just by increasing the timeout, I'll take a look 🤔.

geriux commented:
> Yep, I thought the same, in fact, I tried to solve it by adding an extra wait in [a70d9c7](https://github.com/wordpress-mobile/gutenberg-mobile/commit/a70d9c7a8ea2508989ae49d143fa8cc871fe7021). But still, it didn't work. Maybe it can be fixed just by increasing the timeout, I'll take a look 🤔.

So sorry you're experiencing this flaky test 😓 what if we crop the screenshot? We just need to check that the caret selection has the right color for this test so maybe if we do something like:

```diff --git a/__device-tests__/gutenberg-editor-sanity-test-4-visual.test.js b/__device-tests__/gutenberg-editor-sanity-test-4-visual.test.js
index 56397746f..2cf3eef29 100644
--- a/__device-tests__/gutenberg-editor-sanity-test-4-visual.test.js
+++ b/__device-tests__/gutenberg-editor-sanity-test-4-visual.test.js
@@ -149,8 +149,10 @@ describe( 'Gutenberg Editor - Test Suite 4', () => {
 			);
 
 			// Visual test check
+			const location = await buttonBlock.getLocation();
 			const screenshot = await takeScreenshot( {
 				withoutKeyboard: true,
+				topOffset: location.y,
 			} );
 			expect( screenshot ).toMatchImageSnapshot();
```

```diff --git a/__device-tests__/utils.js b/__device-tests__/utils.js
index 111f824b3..9ccb5f804 100644
--- a/__device-tests__/utils.js
+++ b/__device-tests__/utils.js
@@ -11,12 +11,14 @@ const { isAndroid } = e2eUtils;
  * @param {Object}  options                    Options
  * @param {boolean} [options.withoutKeyboard]  Prevents showing the keyboard in the screenshot.
  * @param {number}  [options.heightPercentage] Specify a custom height in percentage.
+ * @param {number}  [options.topOffset]        Specify a custom top offset to crop.
  * @return {Buffer} Sreenshot image.
  */
-export async function takeScreenshot(
-	options = { withoutKeyboard: false, heightPercentage: undefined }
-) {
-	const { withoutKeyboard, heightPercentage } = options;
+export async function takeScreenshot( {
+	withoutKeyboard,
+	heightPercentage,
+	topOffset,
+} = {} ) {
 	const iPadDevice = process.env.IPAD;
 	const orientation = await editorPage.driver.getOrientation();
 	const isPortrait = orientation === 'PORTRAIT';
@@ -27,6 +29,7 @@ export async function takeScreenshot(
 	const portraitWidthSize = iPadDevice ? 768 : 320;
 	const landscapeWidthSize = iPadDevice ? 1024 : 640;
 	const widthSize = isPortrait ? portraitWidthSize : landscapeWidthSize;
+	const pixelRatio = isAndroid() ? 3.5 : 3;
 
 	const base64Image = Buffer.from( screenshot, 'base64' );
 	const image = await jimp.read( base64Image );
@@ -50,6 +53,20 @@ export async function takeScreenshot(
 		}
 	}
 
+	// Custom top offset
+	if ( topOffset ) {
+		const topOffsetWithRatio = isAndroid()
+			? topOffset
+			: topOffset * pixelRatio;
+		const finalTopOffset = topOffsetWithRatio - statusBarHeight;
+		image.crop(
+			0,
+			finalTopOffset,
+			image.getWidth(),
+			image.getHeight() - finalTopOffset
+		);
+	}
+
 	// Custom height in percentage
 	if ( heightPercentage ) {
 		const height = ( heightPercentage * image.getHeight() ) / 100;
```

It would generate:

iOS|Android
-|-
<kbd><img src="https://github.com/wordpress-mobile/gutenberg-mobile/assets/4885740/7fd6a8c0-60da-40dd-ae0e-a41714f969f5" /></kbd>|<kbd><img src="https://github.com/wordpress-mobile/gutenberg-mobile/assets/4885740/8365c048-89b3-4e5f-a908-32812bfd6d5f" /></kbd>

This would solve the issue of not having the same options in the contextual menu or language. If not, we could disable the test and re-think its approach.

fluiddot commented:
> So sorry you're experiencing this flaky test 😓 what if we crop the screenshot? We just need to check that the caret selection has the right color for this test so maybe if we do something like:

That's a good idea @geriux ! I'd like to explore first why we are getting a different screenshot, but I agree that the approach of cropping the screenshot would definitely solve the problem. Besides, it would be helpful in other cases where we only want to visual test a specific area of the editor. This approach probably would help make the visual testing less flaky.

I haven't being able to focus on moving this forward, but I'll try to spare time today to wrap it up 👍.

fluiddot commented:
> That's a good idea @geriux ! I'd like to explore first why we are getting a different screenshot, but I agree that the approach of cropping the screenshot would definitely solve the problem. Besides, it would be helpful in other cases where we only want to visual test a specific area of the editor. This approach probably would help make the visual testing less flaky.
> 
> I haven't being able to focus on moving this forward, but I'll try to spare time today to wrap it up 👍.

I've checked the video capture in SauceLabs and noticed that the contextual menu is translucid for some reason, which diverges from the local results. That said, I'll take the approach you shared Gerardo, and crop the screenshot to avoid conflicts with that menu between CI and the local environment.

fluiddot committed:
Merge branch 'trunk' into update/test-description-functionality-test-suite-4

fluiddot committed:
Update screenshot in Buttons block e2e test

fluiddot commented:
Following @geriux idea, I've updated the test that was failing using the new helper `takeScreenshotByElement`. You can see how the screen looks now in https://github.com/wordpress-mobile/gutenberg-mobile/pull/6013/commits/f5626013205bf31eba2bfa9057a13e86e01c1fdb. I hope this workaround finally addresses the failures 🤞 .

fluiddot commented:
> Following @geriux idea, I've updated the test that was failing using the new helper `takeScreenshotByElement`. You can see how the screen looks now in [f562601](https://github.com/wordpress-mobile/gutenberg-mobile/commit/f5626013205bf31eba2bfa9057a13e86e01c1fdb). I hope this workaround finally addresses the failures 🤞 .

It's still failing due to just one pixel 🤦. I'll crop the screenshot to avoid including the contextual menu.
 
![image](https://github.com/wordpress-mobile/gutenberg-mobile/assets/14905380/d1dccef1-e981-4ce9-8bbe-5bddac1e55e6)

fluiddot committed:
Update screenshot in Buttons block e2e test

The screenshot is now taken using the Button block instead of the Buttons block.

fluiddot commented:
Finally, all E2E tests passed with the latest fix (https://github.com/wordpress-mobile/gutenberg-mobile/pull/6013/commits/861fa4523c7eb12cb8f6a67f3148c3f15552a8f0) 🎊. I'll proceed now to merge the PR, sorry @geriux and @dcalhoun for the extended delay to finish and merge the PR 🙇 .
